File: 1_dfs_bfs/김민주/게임맵최단거리.cpp
#include <vector>
#include <stdio.h>
#include <iostream>
#include <queue>
using namespace std;

int mvi[] = {0,1,0,-1};
int mvj[] = {1,0,-1,0};

int solution(vector<vector<int> > maps)
{
    int answer = 0;
    int nowi = 0;
    int nowj = 0;
    int n = maps.size();
    int m = maps[0].size();
    queue <pair<int, int>> q;
    int visited[101][101] = {0,};
    q.push({0,0});
    visited[0][0] = 1;
    
    while (!q.empty()){
        nowi = q.front().first;
        nowj = q.front().second;
        q.pop();
        
        if ((nowi == n-1)&&(nowj == m-1)){
            return visited[nowi][nowj];
        }
        
        for (int k=0; k<4; k++){
            int newi = nowi + mvi[k];
            int newj = nowj + mvj[k];
            if ((newi >= 0 && newi < n)&&(newj >= 0 && newj < m)){
                if ((maps[newi][newj] == 1)&&(visited[newi][newj] == 0)){
                    q.push({newi, newj});
                    visited[newi][newj] = visited[nowi][nowj] +1;
                }
            }
        }        
    }
    return -1;
}

File: 1_dfs_bfs/이준우/게임 맵 최단거리.py
from collections import deque

dx = [-1, 0, 1, 0]
dy = [0, -1, 0, 1]


def solution(maps):
    # 그래프의 크기
    n = len(maps)
    m = len(maps[0])

    # 방문하지 않은 경우 -1, 그 외에는 방문한 경우
    visited = [[-1] * m for _ in range(n)]
    
    # 출발지
    q = deque([(0, 0)])
    visited[0][0] = 1
    while q: 
        x, y = q.popleft()
        
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 <= nx < n and 0 <= ny < m and visited[nx][ny] == -1 and maps[nx][ny] == 1:
                if nx == (n - 1) and ny == (m - 1):
                    return visited[x][y] + 1

                q.append((nx, ny))
                visited[nx][ny] = visited[x][y] + 1
                
    
    return -1
